src/service/AccountService.ts

```diff
@@ -0,0 +1,137 @@
+     * @param addresses Array of addresses
+     */
+    public accountInfoWithResolvedMosaic(addresses: Address[]): Observable<AccountInfoResolvedMosaic[]> {
+        const accountInfoObservable = this.accountRepository.getAccountsInfo(addresses);
+        const distinctNames = accountInfoObservable.pipe(
+            mergeMap((info) => {
+                const namespaceIds = this.getDistinctNamespaceIdFromAccountInfors(info);
+                if (namespaceIds.length) {
+                    return this.namespaceRepository.getNamespacesName(namespaceIds);
+                }
+                return of([]);
+            }),
+        );
+
+        return accountInfoObservable.pipe(
+            withLatestFrom(distinctNames),
+            map(([infos, names]) => {
+                return infos.map((info) => {
+                    const resolved = this.resolveMosaics(info.mosaics, names);
+                    return DtoMapping.assign(info, { resolvedMosaics: resolved });
+                });
+            }),
+        );
+    }
+
+    /**
+     * Get namespace info for account with namespace name
+     * @param addresses Array of addresses
+     * @returns {Observable<NamespaceInfoWithName[]>}
+     */
+    public accountNamespacesWithName(addresses: Address[]): Observable<NamespaceInfoWithName[]> {
+        return this.namespaceRepository.getNamespacesFromAccounts(addresses).pipe(
+            mergeMap((infos) => {
+                const namespaceIds = infos.map((i) => i.id);
+                return this.namespaceRepository.getNamespacesName(namespaceIds).pipe(
+                    map((resolved) => {
+                        return infos.map((info) => {
+                            const name = resolved.find((r) => r.namespaceId.equals(info.id));
+                            return DtoMapping.assign(info, { namespaceName: name?.name });
+                        });
+                    }),
+                );
+            }),
+        );
+    }
+
+    /**
+     * Resolve mosaics provides namespace names
+     * @param mosaics unresolved mosaics
+     * @return {ResolvedMosaic[]}
+     */
+    private resolveMosaics(mosaics: Mosaic[], names: NamespaceName[]): ResolvedMosaic[] {
+        return mosaics.map((mosaic) => {
+            if (mosaic.id instanceof MosaicId) {
+                return mosaic as ResolvedMosaic;
+            } else {
+                const name = names.find((f) => f.namespaceId.equals(mosaic.id));
+                if (name) {
+                    return DtoMapping.assign(mosaic, { namespaceName: name });
+                } else {
+                    return mosaic as ResolvedMosaic;
+                }
+            }
+        });
+    }
+
+    /**
+     * Get distince list of namespacesIds from list of accountInfos
+     * @param accountInfo List of account info
+     * @returns {NamespaceId[]}
+     */
+    private getDistinctNamespaceIdFromAccountInfors(accountInfo: AccountInfo[]): NamespaceId[] {
```

Comment:
I'm not sure to understand what this function does, could you please give me some context where could be used?

Reply:
To reduce the rest calls. Get a distinct list of namespaceIds and with 1 rest call to get all names. Then map the resolved names back to models.

Reply:
I see, I haven't seen that's defined as private 👍